### DISCUSSION TITLE
LDM notes for August 24th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#6412] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-08-24.md

## Agenda

1. C# feature triage.

**Comment by huoyaoyuan** [comment]
I feel #6308 is an important complement to generic math feature. A lot of code can't be changed to generic math now. This may require runtime change though.

> Community-driven features

I implemented a prototype of `await?`, and the feature was marked as 'implementation required', but there was no follow-up action.

**Reply by 333fred** [reply]
As far as I can tell, the only mention you've ever made of this is one time in a github issue, and there has been no follow up with any member of the team since.

As you know, we are on discord as well as on github, feel free to follow up with us on this.

**Comment by alrz** [comment]
> 6. Smaller nits,

We'll have to wait for the next round of triage but there's a ton of features already in the Working Set milestone, hopefully at least some of them fits into this category (and possibly some will be pushed out to the backlog? - there's so many it's not clear what the actual status is on those features).

**Reply by PathogenDavid** [reply]
I feel like if working set items are going to degrade they should degrade to either the any time milestone or a new milestone entirely. The backlog and the working set signal two very different levels of interest from the LDT to me.

**Reply by alrz** [reply]
Thinking about what it actually may look like, a lot of languages bucket features into a number of "stages" depending on the design progress. That may be a good idea to explore.

**Comment by MgSam** [comment]
Is there some reason the [pipe operator](https://github.com/dotnet/csharplang/discussions/96) and [switch on System.Type](https://github.com/dotnet/csharplang/discussions/356) never get discussed for prioritization despite being among the top voted discussions in this repo?

Maybe I'm being naïve but they seem to have a good effort / impact ratio.

**Reply by CyrusNajmabadi** [reply]
Neither are championed as something an LDM member thinks is particularly important to do.

**Reply by foxesknow** [reply]
> Is there some reason the [pipe operator](https://github.com/dotnet/csharplang/discussions/96) and [switch on System.Type](https://github.com/dotnet/csharplang/discussions/356) never get discussed for prioritization despite being among the top voted discussions in this repo?
> 

They're too busy adding pointless features like expression bodied members, badly thought out features like the `required` keyword or making a mess of `struct`s

**Reply by HaloFour** [reply]
"Pointless" is subjective and everyone has a different opinion regarding which features should be added to the language and when, including within the language design group.

What is pointless is denigrating the language team for not having opinions that align with your own.

**Reply by foxesknow** [reply]
> "Pointless" is subjective and everyone has a different opinion regarding which features should be added to the language and when, including within the language design group.
> 
> What is pointless is denigrating the language team for not having opinions that align with your own.

Sorry, I didn't realize the language team were above criticism 🙄 

I very much get the impression that the language team live in their own little bubble where they mutually pat themselves on the back for coming up with a new "feature" (yay, top level statements). It's like their haunted by the slow evolution of C++ and therefore feel obliged to constantly churn out various half-baked features in case the language becomes irrelevant. You've only got to look at the comment on [Bill Wagner's](https://devblogs.microsoft.com/dotnet/csharp-11-preview-august-update/) around the `required` feature to see some of the frustrations people have.

Plus I think you'll find it only denigrating if the criticism is unfair, which it isn't.

**Reply by spydacarnage** [reply]
There are acceptable ways to provide constructive, critical feedback.

Please note, your messages are not.

**Reply by HaloFour** [reply]
> Sorry, I didn't realize the language team were above criticism

I am not on the language team.  I do believe they invite criticism, but "pointless" and "badly thought out" are not constructive.

> I very much get the impression that the language team live in their own little bubble where they mutually pat themselves on the back for coming up with a new "feature" (yay, top level statements).

All of these conversations are happening out in the public here and on Discord/Gitter.  I also disagree with numerous decisions that have been ultimately made by the team but when asked they generally always have a well thought out reason for the path that they've taken which does incorporate the feedback they've received through these and many other channels available to them.  I think the team has been pretty open and receptive to community feedback, but that doesn't mean that design decisions are driven by those opinions in aggregate.

**Comment by smoothdeveloper** [comment]
> C#-adjacent tooling experience

To me, it is still surprising to see C# <-> F# navigation is still a miss in Visual Studio, while it works well in Jetbrains Rider.

Is F# adjacent enough in this context for expecting this to evolve in Visual Studio?

**Reply by HaloFour** [reply]
I'd be curious as to why this is.  I assume that both compilers would surface enough information about the members to VS to facilitate this cross-language navigation?  Maybe the F# compiler isn't surfacing the same information?  IIRC JetBrains tools generally use their own in-house compilers/analyzers to drive a lot of their functionality and if they have their own F# analysis that might be why their tools interop better.

**Reply by alrz** [reply]
JB uses a common base for most if not all languages while Roslyn supports exactly two. I'm not familiar with LSP but that may help to unify the experience.

**Reply by HaloFour** [reply]
IIRC Roslyn isn't limited to C# and VB.NET, right?  Those are just the only two languages to have been built on it, and it would likely take a _lot_ of effort to port the existing F# compiler on top of Roslyn.  But that said, aren't the language services in VS a different abstraction in general, which Roslyn and F# would support separately?  I recall being able to navigate cross-language solutions long before Roslyn, and between other languages than C# and VB.NET.

**Reply by En3Tho** [reply]
As far as I can remember there was an attempt to build F# compiler on top of Roslyn but it seemed to be too too too much of effort as F# compiler was rewritten to F# much earlier than C# migrated from C++ to Roslyn 

There is https://github.com/fsprojects/Roslyn.FSharp but it seems to be dead?

**Reply by alrz** [reply]
>LSP

Yeah, this is actually already in progress https://github.com/dotnet/roslyn/pull/62924

**Reply by 333fred** [reply]
That pr is not what you're looking for. That's generalized support code for reading and writing json streams.

**Reply by alrz** [reply]
My understanding from the description was that this is a LSP "framework" for other languages to implement, if not then I was wrong.

**Reply by smoothdeveloper** [reply]
Interestingly, the F# to C# "Go To Definition" scenario already works.

The infrastructure to make this work cross dotnet languages should ideally be something which is not tied to Roslyn nor F# or any language AST, but on resolved CLR representation of elements being pointed/referenced in the projects (based on assemblies with signatures only?), and passing it off the correct language service to resolve the precise location.

I don't believe it is a requirement for C# project system to know about F# AST if it was made like this, and also that it would work with other languages that target CLR and implement the same resolution infrastructure.

Once the navigation is fixed, it would be good for the refactoring use cases that are shared, to work across languages as well.

This would help with any "C#-adjacent" CLR languages that would leverage this infrastructure, and I think, would be a good thing for Visual Studio.
